File: .claude/hooks/setup-check-deps.mjs
#!/usr/bin/env node
// Vérifie que les dépendances du projet sont à jour au démarrage de session (SessionStart)
import { existsSync, statSync } from 'fs';
import { join } from 'path';
import { fileURLToPath } from 'url';

const SPECS = [
  ['pnpm-lock.yaml', 'node_modules', 'pnpm install'],
  ['package-lock.json', 'node_modules', 'npm ci'],
  ['yarn.lock', 'node_modules', 'yarn install --frozen-lockfile'],
  ['requirements.txt', '.venv', 'pip install -r requirements.txt'],
];

export function run({
  exists = existsSync,
  stat = statSync,
  projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
} = {}) {
  const warnings = [];

  for (const [lockfile, modulesDir, installCmd] of SPECS) {
    const lock = join(projectDir, lockfile);
    const mods = join(projectDir, modulesDir);
    if (!exists(lock)) continue;
    if (!exists(mods)) {
      warnings.push(`[setup-check-deps] ⚠ ${modulesDir} absent — lancez : ${installCmd}\n`);
      continue;
    }
    if (stat(lock).mtimeMs > stat(mods).mtimeMs) {
      warnings.push(`[setup-check-deps] ⚠ ${lockfile} plus récent que ${modulesDir} — lancez : ${installCmd}\n`);
    }
  }

  return { warnings, message: warnings.join('') };
}

/* v8 ignore next 4 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result.message) process.stderr.write(result.message);
}

File: .claude/hooks/setup-check-deps.py
#!/usr/bin/env python3
# @hookstack setup-check-install-deps
"""Checks that project dependencies are up to date at session start (SessionStart)."""
import os
import sys

SPECS = [
    ("pnpm-lock.yaml", "node_modules", "pnpm install"),
    ("package-lock.json", "node_modules", "npm ci"),
    ("yarn.lock", "node_modules", "yarn install --frozen-lockfile"),
    # Projet uv : le lockfile uv.lock est la référence (requirements.txt n'est pas un lockfile).
    ("uv.lock", ".venv", "uv sync"),
]


def run(*, exists=None, stat=None, project_dir=None):
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd()
    if exists is None:
        exists = lambda f: os.path.exists(os.path.join(project_dir, f))
    if stat is None:
        stat = os.stat

    warnings = []
    for lockfile, modules_dir, install_cmd in SPECS:
        lock = os.path.join(project_dir, lockfile)
        mods = os.path.join(project_dir, modules_dir)
        if not exists(lock):
            continue
        if not exists(mods):
            warnings.append(
                f"[setup-check-deps] ⚠ {modules_dir} absent — lancez : {install_cmd}\n"
            )
            continue
        if stat(lock).st_mtime > stat(mods).st_mtime:
            warnings.append(
                f"[setup-check-deps] ⚠ {lockfile} plus récent que {modules_dir} — lancez : {install_cmd}\n"
            )

    return {"warnings": warnings, "message": "".join(warnings)}


if __name__ == "__main__":
    result = run()
    if result["message"]:
        sys.stderr.write(result["message"])
